src/cswinrt/strings/WinRT.cs

```diff
@@ -415,63 +421,165 @@ protected virtual System.Delegate EventInvoke
+            {
+                // If cache exists, use it, else create new
+                if (states.ContainsKey(index))
+                {
+                    source._state = states[index];
+                }
+                else
+                {
+                    source._state = new EventSource<TDelegate>.State();
+                    states[index] = source._state;
+                }
+            }
+
+            public static void Create(IObjectReference obj, EventSource<TDelegate> source, int index)
+            {
+                // If event source implements weak reference support, track event registrations so that
+                // unsubscribes will work across garbage collections.  Note that most static/factory classes
+                // do not implement IWeakReferenceSource, so static codegen caching approach is also used.
+                IWeakReference target = null;
+                try
+                {
+                    target = obj.AsInterface<IWeakReferenceSource>().GetWeakReference();
+                }
+                catch (Exception)
+                {
+                    source._state = new EventSource<TDelegate>.State();
+                    return;
+                }
+
+                cachesLock.EnterReadLock();
+                try
+                {
+                    caches.AddOrUpdate(obj.ThisPtr,
+                        (IntPtr ThisPtr) => new Cache(target, source, index),
+                        (IntPtr ThisPtr, Cache cache) => cache.Update(target, source, index));
+                }
+                finally
+                {
+                    cachesLock.ExitReadLock();
+                }
+            }
+
+            public static void Remove(IntPtr thisPtr, int index)
+            {
+                if (caches.TryGetValue(thisPtr, out var cache))
+                {
+                    cache.states.TryRemove(index, out var _);
+                    // using double-checked lock idiom
+                    if (cache.states.IsEmpty)
+                    {
+                        cachesLock.EnterWriteLock();
+                        try
+                        {
+                            if (cache.states.IsEmpty)
+                            {
+                                caches.TryRemove(thisPtr, out var _);
+                            }
+                        }
+                        finally
+                        {
+                            cachesLock.ExitWriteLock();
+                        }
+                    }
+                }
             }
         }
 
         internal EventSource(IObjectReference obj,
             delegate* unmanaged[Stdcall]<System.IntPtr, System.IntPtr, out WinRT.EventRegistrationToken, int> addHandler,
-            delegate* unmanaged[Stdcall]<System.IntPtr, WinRT.EventRegistrationToken, int> removeHandler)
+            delegate* unmanaged[Stdcall]<System.IntPtr, WinRT.EventRegistrationToken, int> removeHandler,
+            int index = 0)
```

Comment:
is index always 0 or am I missing something.  Don't seem to see a place where we pass a value for it.

Reply:
you're not missing anything - dropped out in a merge - good catch